File: auth/password.go
package auth

import (
	"crypto/rand"
	"crypto/subtle"
	"encoding/base64"
	"errors"
	"fmt"
	"strings"

	"golang.org/x/crypto/argon2"
)

const (
	minPasswordLength = 8
	maxPasswordLength = 128
	argonTime         = 3
	argonMemory       = 64 * 1024
	argonThreads      = 4
	argonKeyLength    = 32
	saltLength        = 16
)

var errInvalidPasswordHash = errors.New("invalid password hash")

// HashPassword derives an encoded Argon2id credential hash. The caller must
// discard the plaintext password as soon as this method returns.
func HashPassword(password string) (string, error) {
	if !validPassword(password) {
		return "", ErrInvalidInput
	}

	salt := make([]byte, saltLength)
	if _, err := rand.Read(salt); err != nil {
		return "", fmt.Errorf("generate password salt: %w", err)
	}
	defer clear(salt)

	derived := argon2.IDKey([]byte(password), salt, argonTime, argonMemory, argonThreads, argonKeyLength)
	defer clear(derived)
	return fmt.Sprintf("$argon2id$v=19$m=%d,t=%d,p=%d$%s$%s",
		argonMemory, argonTime, argonThreads,
		base64.RawStdEncoding.EncodeToString(salt),
		base64.RawStdEncoding.EncodeToString(derived)), nil
}

// VerifyPassword compares a plaintext password with an encoded Argon2id hash.
// Malformed stored hashes deliberately compare as invalid credentials.
func VerifyPassword(encoded, password string) bool {
	memory, timeCost, threads, salt, expected, err := parsePasswordHash(encoded)
	if err != nil || !validPassword(password) {
		return false
	}
	defer clear(salt)
	defer clear(expected)

	actual := argon2.IDKey([]byte(password), salt, timeCost, memory, threads, uint32(len(expected)))
	defer clear(actual)
	return subtle.ConstantTimeCompare(actual, expected) == 1
}

func validPassword(password string) bool {
	length := len([]rune(password))
	return length >= minPasswordLength && length <= maxPasswordLength
}

func parsePasswordHash(encoded string) (uint32, uint32, uint8, []byte, []byte, error) {
	parts := strings.Split(encoded, "$")
	if len(parts) != 6 || parts[0] != "" || parts[1] != "argon2id" || parts[2] != "v=19" {
		return 0, 0, 0, nil, nil, errInvalidPasswordHash
	}

	var memory, timeCost uint32
	var threads uint8
	if _, err := fmt.Sscanf(parts[3], "m=%d,t=%d,p=%d", &memory, &timeCost, &threads); err != nil ||
		memory != argonMemory || timeCost != argonTime || threads != argonThreads {
		return 0, 0, 0, nil, nil, errInvalidPasswordHash
	}
	salt, err := base64.RawStdEncoding.Strict().DecodeString(parts[4])
	if err != nil || len(salt) != saltLength {
		clear(salt)
		return 0, 0, 0, nil, nil, errInvalidPasswordHash
	}
	expected, err := base64.RawStdEncoding.Strict().DecodeString(parts[5])
	if err != nil || len(expected) != argonKeyLength {
		clear(salt)
		clear(expected)
		return 0, 0, 0, nil, nil, errInvalidPasswordHash
	}
	return memory, timeCost, threads, salt, expected, nil
}

File: server/src/auth/password.ts
/** Argon2id password hashing, byte-compatible with the Go server's hashes. */

import { argon2id } from "hash-wasm";
import { newSalt } from "../utils/ids.ts";
import { base64UrlEncodeBytes, timingSafeEqual } from "../utils/crypto.ts";
import { TidekeepersError, unauthorized } from "../utils/errors.ts";

export const minPasswordLength = 8;
export const maxPasswordLength = 128;

const argonTime = 3;
const argonMemory = 64 * 1024;
const argonThreads = 4;
const argonKeyLength = 32;
const saltLength = 16;

const encodedPattern =
  /^\$argon2id\$v=19\$m=65536,t=3,p=4\$([A-Za-z0-9_-]+)\$([A-Za-z0-9_-]+)$/;

function validPassword(password: string): boolean {
  const length = Array.from(password).length;
  return length >= minPasswordLength && length <= maxPasswordLength;
}

/**
 * Derives an encoded Argon2id credential hash matching the Go format
 * `$argon2id$v=19$m=65536,t=3,p=4$<salt b64url>$<key b64url>`.
 */
export async function hashPassword(password: string): Promise<string> {
  if (!validPassword(password)) {
    throw new TidekeepersError(
      "invalid authentication input",
      "invalid_request",
      { code: "INVALID_AUTH_INPUT" },
    );
  }
  const salt = newSalt();
  const derived = await argon2id({
    password,
    salt,
    parallelism: argonThreads,
    iterations: argonTime,
    memorySize: argonMemory,
    hashLength: argonKeyLength,
    outputType: "binary",
  });
  return `$${
    [
      "argon2id",
      "v=19",
      `m=${argonMemory},t=${argonTime},p=${argonThreads}`,
      base64UrlEncodeBytes(salt),
      base64UrlEncodeBytes(derived),
    ].join("$")
  }`;
}

/** Compares a plaintext password with an encoded Argon2id hash. */
export async function verifyPassword(
  encoded: string,
  password: string,
): Promise<boolean> {
  const match = encodedPattern.exec(encoded);
  if (!match || !validPassword(password)) {
    return false;
  }
  const salt = base64UrlDecode(match[1]!);
  const expected = base64UrlDecode(match[2]!);
  if (salt.length !== saltLength || expected.length !== argonKeyLength) {
    return false;
  }
  const actual = await argon2id({
    password,
    salt,
    parallelism: argonThreads,
    iterations: argonTime,
    memorySize: argonMemory,
    hashLength: expected.length,
    outputType: "binary",
  });
  return timingSafeEqual(actual, expected);
}

export function invalidCredentials(): TidekeepersError {
  return unauthorized("invalid credentials", { code: "INVALID_CREDENTIALS" });
}

function base64UrlDecode(value: string): Uint8Array {
  const b64 = value.replace(/-/g, "+").replace(/_/g, "/");
  const padded = b64 + "=".repeat((4 - (b64.length % 4)) % 4);
  const binary = atob(padded);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}
